File: 4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

type board struct {
	size    int
	numbers []int
}

var re = regexp.MustCompile("\\d+")

const MARK = -1

func read(scanner *bufio.Scanner) board {
	scanner.Scan()
	first := re.FindAllString(scanner.Text(), -1)
	result := board{size: len(first), numbers: make([]int, len(first)*len(first))}
	for c, s := range first {
		n, _ := strconv.ParseInt(s, 0, 0)
		result.numbers[c] = int(n)
	}
	for r := 1; r < result.size; r++ {
		scanner.Scan()
		for c, s := range re.FindAllString(scanner.Text(), -1) {
			n, _ := strconv.ParseInt(s, 0, 0)
			result.numbers[r*result.size+c] = int(n)
		}
	}
	return result
}

func (b board) String() string {
	return fmt.Sprintf("%v: %v", b.size, b.numbers)
}

func (b board) mark(n int) {
	for i, e := range b.numbers {
		if e == n {
			b.numbers[i] = MARK
		}
	}
}

func (b board) score(n int) int {
	sum := 0
	for _, el := range b.numbers {
		if el != MARK {
			sum += el
		}
	}
	return sum * n
}

func (b board) win() bool {
	for i := 0; i < b.size; i++ {
		marked := true
		for r := 0; (r < b.size) && marked; r++ {
			if b.numbers[r*b.size+i] != MARK {
				marked = false
			}
		}
		if marked {
			return true
		}
		marked = true
		for c := 0; (c < b.size) && marked; c++ {
			if b.numbers[i*b.size+c] != MARK {
				marked = false
			}
		}
		if marked {
			return true
		}
	}
	return false
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	numbers := make([]int, 0, 100)
	for _, s := range re.FindAllString(scanner.Text(), -1) {
		n, _ := strconv.ParseInt(s, 0, 0)
		numbers = append(numbers, int(n))
	}
	boards := make([]board, 0, 100)
	for scanner.Scan() {
		boards = append(boards, read(scanner))
	}
	var first, last, wins int
	for _, n := range numbers {
		for _, b := range boards {
			if b.win() {
				continue
			}
			b.mark(n)
			if b.win() {
				wins++
				if wins == 1 {
					first = b.score(n)
				} else {
					last = b.score(n)
				}
			}
		}
	}
	fmt.Println(first)
	fmt.Println(last)
}

File: 4.rb
#!ruby
class Board
    def initialize(numbers)
        @numbers = numbers
    end

    def self.parse(input)
        numbers = []
        5.times do 
            numbers.append(input.gets.strip.split(" ").map(&:to_i))
        end
        return Board.new(numbers)
    end

    def mark(n)
        @numbers.each_index do |row|
            @numbers[row].each_index do |col|
                if @numbers[row][col] == n then @numbers[row][col] = -1; end
            end
        end
    end

    def win?
        5.times do |i|
            return true if @numbers[i].all? do |n| n == -1 end
            return true if @numbers.all? do |row| row[i] == -1 end
        end
        return false
    end

    def score(n)
        return n * (@numbers.map do |row| row.sum do |i| [0,i].max end end.inject(&:+))
    end
end
numbers = $stdin.gets.strip.split(",").map(&:to_i)
boards = []
while $stdin.gets != nil do boards.append(Board.parse($stdin)) end
score = -1
numbers.each do |n|
    boards.each do |board|
        next if board.win?
        board.mark(n)
        if board.win?
            if score == -1 then puts(board.score(n)) end
            score = board.score(n)
        end
    end 
end
puts(score)